[WebXR][WebGPU] Layers are coupled to WebGPU

svillar opened the pull request:
#### 8e097522b28fefa2ea5eb4fd2ab30148971b49d3
<pre>
[WebXR][WebGPU] Layers are coupled to WebGPU
<a href="https://bugs.webkit.org/show_bug.cgi?id=300241">https://bugs.webkit.org/show_bug.cgi?id=300241</a>

Reviewed by Mike Wyrzykowski.

It should be possible to use WebXR Layers without the WebXR WebGPU bindings.
The current code assumes that WebXR layers can only be used if WebGPU is
available which is not true. This is the first step towards having a
WebGPU-less implementation of WebXR Layers.

No new tests as this is just decoupling the build of different features.

* Source/WebCore/Modules/webxr/XRGPUBinding.cpp:
* Source/WebCore/Modules/webxr/XRGPUBinding.h:
* Source/WebCore/Modules/webxr/XRGPUBinding.idl:
* Source/WebCore/Modules/webxr/XRGPULayerInit.h:
* Source/WebCore/Modules/webxr/XRGPULayerInit.idl:
* Source/WebCore/Modules/webxr/XRGPUProjectionLayerInit.h:
* Source/WebCore/Modules/webxr/XRGPUProjectionLayerInit.idl:
* Source/WebCore/Modules/webxr/XRGPUSubImage.cpp:
* Source/WebCore/Modules/webxr/XRGPUSubImage.h:
* Source/WebCore/Modules/webxr/XRGPUSubImage.idl:
* Source/WebCore/Modules/webxr/XRProjectionLayer.cpp:
(WebCore::XRProjectionLayer::startFrame):
(WebCore::XRProjectionLayer::endFrame):
(WebCore::XRProjectionLayer::textureWidth const):
(WebCore::XRProjectionLayer::textureHeight const):
(WebCore::XRProjectionLayer::textureArrayLength const):
* Source/WebCore/Modules/webxr/XRProjectionLayer.h:

Canonical link: <a href="https://commits.webkit.org/301196@main">https://commits.webkit.org/301196@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f0df861e41eff400d2dc40a32c7a66969727c167

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/125093 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/44760 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/35499 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/131941 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/77076 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/126970 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/45451 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/53325 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/95226 "25 flakes 29 failures") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/63262 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/128047 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/36280 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/111875 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/75768 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/35190 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/30035 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/75422 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/117184 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/106045 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/30264 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/134620 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/123611 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/51907 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/39702 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/103693 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/52337 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/108092 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/103465 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26373 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/48816 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/27107 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/48960 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/51796 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/57586 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/156634 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/51166 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/39230 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/54522 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/52857 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/d92375ac0d117209f402cd1ee6675cf517b70f21. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51874

mwyrzykowski commented on Source/WebCore/Modules/webxr/XRGPUBinding.idl:
```diff
@@ -26,6 +26,7 @@
 // https://github.com/immersive-web/WebXR-WebGPU-Binding/blob/main/explainer.md
 [
     Conditional=WEBXR_LAYERS,
+    Conditional=WEBGPU,
```
Is this defined anywhere? 

In the other idl files we are using `EnabledBySetting=WebGPUEnabled`

svillar replied:
It's true that there is no default value in `PlatformEnable.h`, we might consider adding it there too, but it should be available for builds.

mwyrzykowski replied:
Adding to PlatformEnable.h would be fine :)

svillar commented:
> `ENABLE_WEBGPU` is not defined anywhere.
> 
> We do have `ENABLE_WEBGPU_BY_DEFAULT` and `HAVE_WEBGPU_IMPLEMENTATION`.
> 
> I don't know if the `HAVE` macros work in the .idl files, so I would propose the `EnabledBySetting=WebGPUEnabled`

For the IDLs that would work but we do need to guard blocks of code as well, so we'd need to export the ENABLE_ to the build I guess

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/f0df861e41eff400d2dc40a32c7a66969727c167. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/51874

svillar committed:
[WebXR][WebGPU] Layers are coupled to WebGPU
https://bugs.webkit.org/show_bug.cgi?id=300241

Reviewed by Mike Wyrzykowski.

It should be possible to use WebXR Layers without the WebXR WebGPU bindings.
The current code assumes that WebXR layers can only be used if WebGPU is
available which is not true. This is the first step towards having a
WebGPU-less implementation of WebXR Layers.

No new tests as this is just decoupling the build of different features.

* Source/WebCore/Modules/webxr/XRGPUBinding.cpp:
* Source/WebCore/Modules/webxr/XRGPUBinding.h:
* Source/WebCore/Modules/webxr/XRGPUBinding.idl:
* Source/WebCore/Modules/webxr/XRGPULayerInit.h:
* Source/WebCore/Modules/webxr/XRGPULayerInit.idl:
* Source/WebCore/Modules/webxr/XRGPUProjectionLayerInit.h:
* Source/WebCore/Modules/webxr/XRGPUProjectionLayerInit.idl:
* Source/WebCore/Modules/webxr/XRGPUSubImage.cpp:
* Source/WebCore/Modules/webxr/XRGPUSubImage.h:
* Source/WebCore/Modules/webxr/XRGPUSubImage.idl:
* Source/WebCore/Modules/webxr/XRProjectionLayer.cpp:
(WebCore::XRProjectionLayer::startFrame):
(WebCore::XRProjectionLayer::endFrame):
(WebCore::XRProjectionLayer::textureWidth const):
(WebCore::XRProjectionLayer::textureHeight const):
(WebCore::XRProjectionLayer::textureArrayLength const):
* Source/WebCore/Modules/webxr/XRProjectionLayer.h:

Canonical link: https://commits.webkit.org/301196@main
